### PR TITLE
Compile tests on PR, to ensure we don't break the build when merging. Alert if we merge and the build fails.

### PR DESCRIPTION
**Pull request by michaelkaye**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [X] Technical
- [ ] Other :

## Content

Run the compilation steps for unit tests and integration tests on each PR, to avoid build tests breaking during nightly run on develop.

Also uses linux runners on the PR branches to speed up time to get a runner (we use macos on the nightly runs for performance running the tests)

Add a notification step that triggers if the build fails on develop or main, and sends a confetti-filled ball to the channel to announce with pleasure that the build has failed and someone should do something about it.

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [X] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

**Comment by michaelkaye** [comment]
The test compilation is expected to fail at the moment as develop contains a issue - so we would expect ` Test / Matrix SDK - Build Android Tests (pull_request) Failing after 5m — Matrix SDK - Build Android Tests ` to be broken.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -41,3 +85,21 @@ jobs:
...
+      - build-android-test-app
+    if: ${{ (github.ref == 'refs/heads/develop' || github.ref == 'refs/heads/main' ) && failure() }}
+    steps:
+      - uses: michaelkaye/matrix-hookshot-action@v0.3.0
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          matrix_access_token: ${{ secrets.ELEMENT_ANDROID_NOTIFICATION_ACCESS_TOKEN }}
+          matrix_room_id: ${{ secrets.ELEMENT_ANDROID_INTERNAL_ROOM_ID }}
+          text_template: "Build is broken for ${{ github.ref }} 🎊: {{#each job_statuses }}{{#with this }}{{#if completed }}  {{name}} {{conclusion}} at {{completed_at}}, {{/if}}{{/with}}{{/each}}"
+          html_template: "Build is broken for ${{ github.ref }} 🎊: {{#each job_statuses }}{{#with this }}{{#if completed }}<br />{{icon conclusion }}{{name}} <font color='{{color conclusion }}'>{{conclusion}} at {{completed_at}} <a href=\"{{html_url}}\">[details]</a></font>{{/if}}{{/with}}{{/each}}"
```
Comment (bmarty) [inline .github/workflows/tests.yml:104]: Emojis FTW! But here we have an error, so maybe use a "error" emoji?
Applicable to line 103 as well

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -12,6 +12,50 @@ env:
...
+      - uses: actions/cache@v2 
+        with: 
+          path: | 
+            ~/.gradle/caches 
+            ~/.gradle/wrapper 
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }} 
+          restore-keys: | 
+            ${{ runner.os }}-gradle- 
+      - name: Build Android Tests for vector 
+        run: ./gradlew clean vector:assembleAndroidTest $CI_GRADLE_ARG_PROPERTIES --stacktrace
```
Comment (bmarty) [inline .github/workflows/tests.yml:57]: We can have (or will have) android tests in other module, so I think it should be better to have only one task and call `./gradlew clean assembleAndroidTest` to build tests for every modules (present and future).

I have tested this command on latest develop and have the same expected result:

```
> Task :matrix-sdk-android:compileDebugAndroidTestKotlin FAILED
e: warnings found and -Werror specified
w: /Users/bmarty/workspaces/element-android/matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt: (293, 65): Name shadowed: keysBackupService
w: /Users/bmarty/workspaces/element-android/matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt: (528, 34): This expression will be resolved to Int in future releases. Please add explicit conversion call
```

**Review by bmarty — CHANGES_REQUESTED** [review]
Thanks. Please see my remarks.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (bmarty) [reply .github/workflows/tests.yml:104]: Ah, my bad, I hadn't read

> Add a notification step that triggers if the build fails on develop or main, and sends a confetti-filled ball to the channel to announce with pleasure that the build has failed and someone should do something about it.

Please ignore me!

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -41,3 +65,20 @@ jobs:
...
         with:
           files: ./**/build/test-results/**/*.xml
+
+# Notify the channel about runs against develop or main that have failures, as PRs should have caught these first.
+  notify:
+    runs-on: ubuntu-latest
+    needs:
+      - unit-tests
+      - build-android-tests
+    if: ${{ (github.ref == 'refs/heads/develop' || github.ref == 'refs/heads/main' ) && failure() }}
```
Comment (michaelkaye) [inline .github/workflows/tests.yml:75]: NB: failure() does not fire if the above tests is cancelled due to concurrency kicking in - this should only run once.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (michaelkaye) [reply .github/workflows/tests.yml]: I think this is fixed in 400a47c

**Review by bmarty — APPROVED** [review]
LGTM! Thanks for the fast update.
(and sorry about the Emoji, next time I will read the description more carefully!)

**Comment by michaelkaye** [comment]
Once merged, we should reenable the required status on the check that the tests compile.

**Pull request merged**